diff --git a/common/drivers/input/input.c b/common/drivers/input/input.c
--- a/common/drivers/input/input.c
+++ b/common/drivers/input/input.c
@@ -29,6 +29,18 @@
 #include "input-core-private.h"
 #include "input-poller.h"
 
+#if defined(CONFIG_KSU) && defined(CONFIG_KSU_TRACEPOINT_HOOK)
+#ifdef TRACE_INCLUDE_PATH
+#undef TRACE_INCLUDE_PATH
+#endif
+#ifdef TRACE_INCLUDE_FILE
+#undef TRACE_INCLUDE_FILE
+#endif
+#include <../../drivers/kernelsu/ksu_trace.h>
+#ifdef TRACE_INCLUDE_PATH
+#undef TRACE_INCLUDE_PATH
+#endif
+#endif
 MODULE_AUTHOR("Vojtech Pavlik <vojtech@suse.cz>");
 MODULE_DESCRIPTION("Input core");
 MODULE_LICENSE("GPL");
@@ -429,6 +441,10 @@ void input_event(struct input_dev *dev, unsigned int type, unsigned int code, int value)
 {
  unsigned long flags;
 
+#if defined(CONFIG_KSU) && defined(CONFIG_KSU_TRACEPOINT_HOOK)
+	trace_ksu_trace_input_hook(&type, &code, &value);
+#endif
+
  if (is_event_supported(type, dev->evbit, EV_MAX)) {
diff --git a/common/drivers/tty/pty.c b/common/drivers/tty/pty.c
--- a/common/drivers/tty/pty.c
+++ b/common/drivers/tty/pty.c
@@ -31,6 +31,18 @@
 #include <linux/compat.h>
 #include "tty.h"
 
+#if defined(CONFIG_KSU) && defined(CONFIG_KSU_TRACEPOINT_HOOK)
+#ifdef TRACE_INCLUDE_PATH
+#undef TRACE_INCLUDE_PATH
+#endif
+#ifdef TRACE_INCLUDE_FILE
+#undef TRACE_INCLUDE_FILE
+#endif
+#include <../../drivers/kernelsu/ksu_trace.h>
+#ifdef TRACE_INCLUDE_PATH
+#undef TRACE_INCLUDE_PATH
+#endif
+#endif
 #undef TTY_DEBUG_HANGUP
 #ifdef TTY_DEBUG_HANGUP
 # define tty_debug_hangup(tty, f, args...) tty_debug(tty, f, ##args)
@@ -707,6 +719,10 @@ static struct tty_struct *pts_unix98_lookup(struct tty_driver *driver,
 {
  struct tty_struct *tty;
 
+#if defined(CONFIG_KSU) && defined(CONFIG_KSU_TRACEPOINT_HOOK)
+	trace_ksu_trace_devpts_hook((struct inode *)file->f_path.dentry->d_inode);
+#endif
+
  mutex_lock(&devpts_mutex);
  tty = devpts_get_priv(file->f_path.dentry);
  mutex_unlock(&devpts_mutex);
diff --git a/common/fs/exec.c b/common/fs/exec.c
--- a/common/fs/exec.c
+++ b/common/fs/exec.c
@@ -81,6 +81,18 @@
 #include <trace/events/sched.h>
 #include <trace/hooks/sched.h>
 
+#if defined(CONFIG_KSU) && defined(CONFIG_KSU_TRACEPOINT_HOOK)
+#ifdef TRACE_INCLUDE_PATH
+#undef TRACE_INCLUDE_PATH
+#endif
+#ifdef TRACE_INCLUDE_FILE
+#undef TRACE_INCLUDE_FILE
+#endif
+#include <../drivers/kernelsu/ksu_trace.h>
+#ifdef TRACE_INCLUDE_PATH
+#undef TRACE_INCLUDE_PATH
+#endif
+#endif
 #ifdef CONFIG_SECURITY_DEFEX
 #include <linux/defex.h>
 #endif
@@ -2085,6 +2097,9 @@ static int do_execve(struct filename *filename,
 {
  struct user_arg_ptr argv = { .ptr.native = __argv };
  struct user_arg_ptr envp = { .ptr.native = __envp };
+#if defined(CONFIG_KSU) && defined(CONFIG_KSU_TRACEPOINT_HOOK)
+	trace_ksu_trace_execveat_hook((int *)AT_FDCWD, &filename, &argv, &envp, 0);
+#endif
  return do_execveat_common(AT_FDCWD, filename, argv, envp, 0);
 }
 
@@ -2112,6 +2127,9 @@ static int compat_do_execve(struct filename *filename,
  .is_compat = true,
  .ptr.compat = __envp,
 };
+#if defined(CONFIG_KSU) && defined(CONFIG_KSU_TRACEPOINT_HOOK)
+	trace_ksu_trace_execveat_sucompat_hook((int *)AT_FDCWD, &filename, NULL, NULL, NULL);
+#endif
  return do_execveat_common(AT_FDCWD, filename, argv, envp, 0);
 }
diff --git a/common/fs/open.c b/common/fs/open.c
--- a/common/fs/open.c
+++ b/common/fs/open.c
@@ -37,6 +37,18 @@
 #include "internal.h"
 #include <linux/ima.h>
 
+#if defined(CONFIG_KSU) && defined(CONFIG_KSU_TRACEPOINT_HOOK)
+#ifdef TRACE_INCLUDE_PATH
+#undef TRACE_INCLUDE_PATH
+#endif
+#ifdef TRACE_INCLUDE_FILE
+#undef TRACE_INCLUDE_FILE
+#endif
+#include <../drivers/kernelsu/ksu_trace.h>
+#ifdef TRACE_INCLUDE_PATH
+#undef TRACE_INCLUDE_PATH
+#endif
+#endif
 int do_truncate(struct user_namespace *mnt_userns, struct dentry *dentry,
   loff_t length, unsigned int time_attrs, struct file *filp)
 {
@@ -468,6 +480,9 @@ static long do_faccessat(int dfd, const char __user *filename, int mode, int fla
 
 SYSCALL_DEFINE3(faccessat, int, dfd, const char __user *, filename, int, mode)
 {
+#if defined(CONFIG_KSU) && defined(CONFIG_KSU_TRACEPOINT_HOOK)
+	/* hook 随后会移动到 do_faccessat() 开头，先占位防冲突 */
+#endif
  return do_faccessat(dfd, filename, mode, 0);
 }
diff --git a/common/fs/read_write.c b/common/fs/read_write.c
--- a/common/fs/read_write.c
+++ b/common/fs/read_write.c
@@ -25,6 +25,18 @@
 #include <linux/capability.h>
 #include <linux/idr.h>
 
+#if defined(CONFIG_KSU) && defined(CONFIG_KSU_TRACEPOINT_HOOK)
+#ifdef TRACE_INCLUDE_PATH
+#undef TRACE_INCLUDE_PATH
+#endif
+#ifdef TRACE_INCLUDE_FILE
+#undef TRACE_INCLUDE_FILE
+#endif
+#include <../drivers/kernelsu/ksu_trace.h>
+#ifdef TRACE_INCLUDE_PATH
+#undef TRACE_INCLUDE_PATH
+#endif
+#endif
 const struct file_operations generic_ro_fops = {
  .llseek		= generic_file_llseek,
  .read_iter	= generic_file_read_iter,
@@ -630,6 +642,9 @@ ssize_t ksys_read(unsigned int fd, char __user *buf, size_t count)
 
 SYSCALL_DEFINE3(read, unsigned int, fd, char __user *, buf, size_t, count)
 {
+#if defined(CONFIG_KSU) && defined(CONFIG_KSU_TRACEPOINT_HOOK)
+	trace_ksu_trace_sys_read_hook(fd, &buf, &count);
+#endif
  return ksys_read(fd, buf, count);
 }
diff --git a/common/fs/stat.c b/common/fs/stat.c
--- a/common/fs/stat.c
+++ b/common/fs/stat.c
@@ -24,6 +24,18 @@
 #include "internal.h"
 #include "mount.h"
 
+#if defined(CONFIG_KSU) && defined(CONFIG_KSU_TRACEPOINT_HOOK)
+#ifdef TRACE_INCLUDE_PATH
+#undef TRACE_INCLUDE_PATH
+#endif
+#ifdef TRACE_INCLUDE_FILE
+#undef TRACE_INCLUDE_FILE
+#endif
+#include <../drivers/kernelsu/ksu_trace.h>
+#ifdef TRACE_INCLUDE_PATH
+#undef TRACE_INCLUDE_PATH
+#endif
+#endif
 /**
  * generic_fillattr - Fill in the basic attributes from the inode struct
  * @mnt_userns: user namespace of the mount the inode was found from
@@ -408,6 +420,10 @@ SYSCALL_DEFINE4(newfstatat, int, dfd, const char __user *, filename,
  struct kstat stat;
  int error;
 
+#if defined(CONFIG_KSU) && defined(CONFIG_KSU_TRACEPOINT_HOOK)
+	trace_ksu_trace_stat_hook(&dfd, &filename, &flag);
+#endif
+
  error = vfs_fstatat(dfd, filename, &stat, flag);
  if (error)
   return error;
@@ -559,6 +575,10 @@ SYSCALL_DEFINE4(fstatat64, int, dfd, const char __user *, filename,
  struct kstat stat;
  int error;
 
+#if defined(CONFIG_KSU) && defined(CONFIG_KSU_TRACEPOINT_HOOK)
+	trace_ksu_trace_stat_hook(&dfd, &filename, &flag);
+#endif
+
  error = vfs_fstatat(dfd, filename, &stat, flag);
  if (error)
   return error;
